linkerd/stack/src/failfast.rs

```diff
@@ -43,112 +105,192 @@ pub enum ResponseFuture<F> {
     FailFast(&'static str),
 }
 
-// === impl FailFast ===
+// === impl Advertise ===
 
-impl<S> FailFast<S> {
-    pub fn layer(
-        scope: &'static str,
-        max_unavailable: Duration,
-    ) -> impl layer::Layer<S, Service = Self> + Clone + Copy {
-        layer::mk(move |inner| Self {
+impl<S> Advertise<S> {
+    fn new(scope: &'static str, inner: S, shared: Arc<Shared>) -> Self {
+        Self {
             scope,
             inner,
-            max_unavailable,
-            // The sleep is reset whenever the service becomes unavailable; this
-            // initial one will never actually be used, so it's okay to start it
-            // now.
-            wait: Box::pin(time::sleep(Duration::default())),
-            state: State::Open,
-        })
+            shared,
+            is_waiting: false,
+            waiting: ReusableBoxFuture::new(async move { unreachable!() }),
+        }
     }
 }
 
-impl<S> Clone for FailFast<S>
+impl<S> Clone for Advertise<S>
 where
     S: Clone,
 {
     fn clone(&self) -> Self {
-        // When cloning failfast, we can't preserve the waiting state, so each
-        // clone will have to detect its own failfast. Practically, this means
-        // that each connection will have to wait for a timeout before
-        // triggering failfast.
-        Self {
-            scope: self.scope,
-            inner: self.inner.clone(),
-            max_unavailable: self.max_unavailable,
+        Self::new(self.scope, self.inner.clone(), self.shared.clone())
+    }
+}
+
+impl<S, T> tower::Service<T> for Advertise<S>
+where
+    S: tower::Service<T>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let scope = self.scope;
+        loop {
+            if self.is_waiting {
+                // We are currently waiting for the inner service to exit failfast.
+                trace!("Advertise: waiting for {scope} service to become ready",);
+                ready!(self.waiting.poll_unpin(cx));
+                trace!("Advertise: {scope} service became ready");
+                self.is_waiting = false;
+            } else if self.shared.in_failfast.load(Ordering::Acquire) {
+                // We are in failfast. start waiting for the inner service to
+                // exit failfast.
+                trace!("Advertise: {scope} service in failfast, waiting for readiness",);
+                let shared = self.shared.clone();
+                self.waiting.set(async move {
+                    shared.notify.notified().await;
+                    trace!("Advertise: {scope} service has become ready");
```

Comment:
We should avoid this scope stuff, imo. We only have it here so that the failfast error messages include some information about which failfast we hit. Better relevant scope information is already going to be in the trace context.

I'd really like to get rid of the `scope` thing entirely`. I think we have two options: (1) At the top of each stack we put a layer that wraps errors with stack context; or (2) we make failfast generic over the type of error it returns. I'd lean to the former...

Anyway, for now, let's omit scope from the logs.

Reply:
okay, i can remove the field entirely here if you like?

Reply:
Yeah I think that's probably good. maybe leave a TODO in the layer about error annotations?